### PR TITLE
feat(solana-driver): Solver interaction

### PR DESCRIPTION
**Pull request by tilacog**
# Description
Adds solver interaction for the `solana-driver` crate.

# Changes
- New `Solver` type that sends `POST /solve` requests to configured solvers. It is now part of the API `State` struct.
- Added a minimal set of `domain` types to support this interaction: `Auction, Order, OrderUID`, `Solution`, etc. 
  - Each with a mininmal set of fields too, because right now they serve just to produce the DTO set for the solver. 
  - I'll expand on them as we integrate w/ the autopilot and with the Settlement program.
- Added a set of DTO definitions to support this interaction, mirroring the `solana-solvers` DTO for the Jupiter API and Solana SDK types returned by it.
- Small refactor on the `solana-solvers` `axum` API to split `bind` and `serve`, so we can import the solver process and run it in integration tests _(following the workspace convention)_.
- Added a new `URL` deserialization helper fn in `serde-ext` to prevent URLs without a trailing slash to be set in config _(after [this bot review comment](https://github.com/cowprotocol/services/pull/4749#discussion_r3798152950))_

## How to test
New live integration test: `driver_solves_against_live_jupiter_engine`, run it with:
```bash
$ cargo nextest run -p solana-driver --run-ignored ignored-only --test jupiter_live
```

### CODE REVIEW

**Comment by squadgazzz** [comment]
@tilacog , please resolve conflicts. Otherwise, the diff is confusing.

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /// POST the auction to this engine's `/solve` endpoint and return the
+    /// domain solutions it produced.
+    #[tracing::instrument(name = "solver_engine", skip_all, fields(solver = %self.name))]
+    pub async fn solve(&self, auction: &domain::Auction) -> Result<Vec<domain::Solution>, Error> {
+        let auction_dto = Auction::new(auction, self.account);
+        let body = serde_json::to_string(&auction_dto)?;
+
+        let solve_url = self.base_url.join("solve").expect("valid /solve path");
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs:49]: Fixed in d6bb454e4ce37c7a2647ad9357fe8e912852b357 at parse time

File: crates/solana-driver/src/infra/solver/dto/solution.rs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                Ok(domain::Solution {
+                    id: solution.id,
+                    solver,
+                    trades,
+                    interactions: solution.interactions.into_iter().map(Into::into).collect(),
+                    address_lookup_tables: solution.address_lookup_tables,
+                    cu_estimate: solution.cu_estimate,
+                })
+            })
+            .collect()
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/solution.rs:155]: Reworded the doc in 34ff1f3c3abc849dccfb5e3e46105699f6caa1fe.
Behavior is unchanged to match the EVM driver. _(rejects the entire solver response on mismatch)_

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                    solver = %self.name,
+                    "auction deadline exceeded before sending request to solver"
+                );
+                return Ok(Vec::new());
+            }
+            // Safe: we just checked `remaining` is positive.
+            remaining.to_std().unwrap()
+        };
+
+        let _permit = self.in_flight.acquire().await;
```
[the remark this answers is not in the document]
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs]: Fixed in f95991dcb9a7f9de46a5836a327679c835353b8e

**Comment by tilacog** [comment]
@squadgazzz I've reconciled this branch w/ latest `main`, should be good for a review now.

File: crates/solana-driver/src/util/mod.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+/// Derive the associated token account (ATA) address for `owner` and `mint`
+/// under the SPL Token program.
+pub fn associated_token_address(owner: &Pubkey, mint: &Pubkey) -> Pubkey {
+    Pubkey::find_program_address(
+        &[owner.as_ref(), SPL_TOKEN_PROGRAM_ID.as_ref(), mint.as_ref()],
+        &SPL_ASSOCIATED_TOKEN_ACCOUNT_PROGRAM_ID,
+    )
+    .0
+}
```
Comment (squadgazzz) [inline crates/solana-driver/src/util/mod.rs:30]: This hardcodes the classic SPL token program into the ATA seeds. For a token-2022 mint, the derived `buy_destination` is probably wrong, since the seeds take the mint's owning token program, so the solver would route the output to an address that will never exist. 

The indexer already treats both programs as first-class `(TOKEN_PROGRAMS = [spl_token, spl_token_2022])`. Either take the token program as a parameter/resolve it from the mint, or reject non-classic mints loudly until that lands, wdyt?

File: crates/solana-driver/src/infra/solver/dto/mod.rs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+//! Wire DTOs for the driver <-> solver-engine HTTP boundary.
+//!
+//! These structs are the driver's own mirror of the `solana-solvers`
+//! `dto/{auction,solution}.rs` types; they are deliberately not shared with the
+//! solver crate so the wire format can evolve on one side at a time. Serde
+//! tests in each module pin the JSON shape against the literals the solver
+//! crate tests assert, so drift breaks a test instead of an auction.
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/mod.rs]: Not sure I am following.  Nothing pins the JSON literals today, so drift breaks an auction instead of a test. Probably worth adding the two shape tests, the doc already promises them.

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        // silently eat into the budget and let the solve run past the deadline.
+        let timeout = {
+            let remaining = auction.deadline.signed_duration_since(chrono::Utc::now());
+            if remaining <= chrono::Duration::zero() {
+                tracing::warn!(
+                    solver = %self.name,
+                    "auction deadline exceeded before sending request to solver"
+                );
+                return Ok(Vec::new());
+            }
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/mod.rs:72]: When the deadline has already passed, this returns `Ok(Vec::new())`, which the caller can't tell apart from "the engine genuinely found nothing". Feels like it should be an `Error::DeadlineExceeded`.

File: crates/solana-driver/src/infra/solver/dto/solution.rs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                    .into_iter()
+                    .map(|trade| {
+                        if !allowed_uids.contains(&trade.order_uid) {
+                            return Err(Error::UnknownOrderUid(trade.order_uid));
+                        }
+                        // The wire format carries no clearing prices or fee yet, so
+                        // the fee defaults to zero.
+                        Ok(domain::Trade {
+                            order_uid: trade.order_uid,
+                            executed_amount: trade.executed_amount,
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/solution.rs:141]: A solver can return `executed_amount > order.amount`. The chain rejects it eventually, but the driver would burn the cycle building and simulating a doomed settlement.

File: crates/solana-driver/src/infra/solver/dto/auction.rs [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    serde_with::serde_as,
+    solana_sdk::pubkey::Pubkey,
+};
+
+/// The auction the driver posts to `/solve`.
+#[serde_as]
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Auction {
+    pub id: u64,
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/auction.rs]: The autopilot request sends i64 (see autopilot-svm/src/infra/driver/dto.rs). If the driver's domain says u64, the two crates disagree on the same field, and integration needs a cast that can theoretically fail. Probably makes sense to use i64 here as well.

File: crates/solana-driver/src/domain/solution.rs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+/// A fulfillment of one auction order.
+#[derive(Clone, Debug)]
+pub struct Trade {
+    pub order_uid: OrderUid,
+    /// Sell-token units for sell orders, buy-token units for buy orders.
+    pub executed_amount: u64,
+    /// Solver-reported fee, in sell-token units. Defaults to zero when the
+    /// wire format carries no fee.
+    pub fee: u64,
```
Comment (squadgazzz) [inline crates/solana-driver/src/domain/solution.rs]: Does it make sense to drop it? Nobody uses it until MVP.

File: crates/solana-driver/src/infra/solver/dto/solution.rs [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    pub accounts: Vec<AccountMetaDto>,
+    #[serde_as(as = "serde_with::base64::Base64")]
+    pub instruction_data: Vec<u8>,
+}
+
+/// Account meta in the driver DTO shape.
+#[serde_as]
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AccountMetaDto {
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/solution.rs]: We are already in the `dto` module, right?
```suggestion
pub struct AccountMeta {
```

File: crates/solana-driver/src/infra/solver/dto/solution.rs [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+    /// Sell-token units for sell orders, buy-token units for buy orders.
+    #[serde_as(as = "serde_with::DisplayFromStr")]
+    pub executed_amount: u64,
+}
+
+/// A Solana instruction the solver supplies, carried verbatim.
+#[serde_as]
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct InstructionDto {
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/solution.rs]: ditto
```suggestion
pub struct Instruction {
```

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+    /// POST the auction to this engine's `/solve` endpoint and return the
+    /// domain solutions it produced.
+    #[tracing::instrument(name = "solver_engine", skip_all, fields(solver = %self.name))]
+    pub async fn solve(&self, auction: &domain::Auction) -> Result<Vec<domain::Solution>, Error> {
+        let auction_dto = Auction::new(auction, self.account);
+        let body = serde_json::to_string(&auction_dto)?;
+
+        let solve_url = self.base_url.join("solve").expect("valid /solve path");
+
+        let _permit = self.in_flight.acquire().await;
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/mod.rs]:
```suggestion
        let _permit = self.in_flight.acquire().await.expect("semaphore is never closed");
```
IIUC, there will be no permit in case or Err, which will allow boundless processing.

File: crates/solana-driver/src/util/mod.rs (hunk 4 above)
Comment (tilacog) [reply crates/solana-driver/src/util/mod.rs:30]: Makes sense, the driver can detect/infer the correct token program by the mint account. I'll add that.

File: crates/serde-ext/src/url.rs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+pub fn deserialize_url_with_trailing_slash<'de, D>(deserializer: D) -> Result<Url, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let mut url = Url::deserialize(deserializer)?;
+    if !url.path().ends_with('/') {
+        url.set_path(&format!("{}/", url.path()));
+    }
+    Ok(url)
+}
```
Comment (squadgazzz) [inline crates/serde-ext/src/url.rs:16]: We probably have this in multiple places already, but adding the evm-heavy `shared` crate dependency doesn't really justify the deduplication.

File: crates/solana-driver/src/infra/solver/dto/mod.rs (hunk 5 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/mod.rs]: That comment was something I left behind by mistake. There was a JSON test I removed because I didn't think it was very useful.

I will bring it back now that your comment made me reconsider it.

File: crates/solana-driver/src/infra/solver/dto/solution.rs (hunk 10 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/solution.rs]: I set it like that to avoid collisions w/ the Solana SDK types, but I'll make the aliasing more uniform here.

File: crates/solana-driver/src/infra/solver/dto/auction.rs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    /// Settlement signer the swap instructions are built for.
+    #[serde_as(as = "serde_with::DisplayFromStr")]
+    pub taker: Pubkey,
+    pub orders: Vec<Order>,
+}
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/dto/auction.rs:27]: We should add the deadline here and on the solana-solvers side I think, like for the EVM to respect it: https://github.com/cowprotocol/services/blob/381aa8a821bdd52dab2eb5f85df4b8d7ab0441de/crates/solvers-dto/src/auction.rs#L25-L25

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+        // Calculate the time remaining until the auction's deadline. This is
+        // computed *after* acquiring the permit, otherwise the wait could
+        // silently eat into the budget and let the solve run past the deadline.
+        let timeout = {
+            let remaining = auction.deadline.signed_duration_since(chrono::Utc::now());
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/mod.rs:65]: https://github.com/cowprotocol/services/blob/31ea719f35072bc6155fdeb990bbc27c9b8833b3/crates/driver/src/domain/time.rs#L31-L31

IIUC, the engine gets the whole remaining budget as its HTTP timeout, so a response landing right before the deadline leaves the driver zero time to build and simulate the settlement. EVM splits this via `solving_share_of_deadline`. Fine to postpone, but a TODO naming the split here would keep it from getting lost.

File: crates/solana-driver/src/domain/auction.rs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+/// One order available for solvers to fill.
+#[derive(Clone, Debug)]
+pub struct Order {
+    pub uid: OrderUid,
+    pub sell_mint: Pubkey,
+    pub buy_mint: Pubkey,
+    /// Sell amount for sells, buy amount for buys.
+    pub amount: u64,
```
Comment (MartinquaXD) [inline crates/solana-driver/src/domain/auction.rs]: Where is the "other" amount of the order? We need to know it to evaluate solutions and compare prices.

File: crates/solana-driver/src/infra/solver/dto/auction.rs [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+/// The auction the driver posts to `/solve`.
+#[serde_as]
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Auction {
+    pub id: u64,
+    /// Settlement signer the swap instructions are built for.
+    #[serde_as(as = "serde_with::DisplayFromStr")]
+    pub taker: Pubkey,
```
Comment (MartinquaXD) [inline crates/solana-driver/src/infra/solver/dto/auction.rs]: It's unclear to me what this is supposed to be. Like what value will this be populated with and why?

File: crates/solana-driver/src/infra/solver/mod.rs (hunk 15 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs:65]: Added a TODO comment in 72f91045e16cd906da4504a8bb0b1dfedcbf12c5

File: crates/solana-driver/src/infra/solver/mod.rs (hunk 12 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs]: Addressed in b6e7ff0edf8344980befbcc0118fa07e21c40dc0

File: crates/solana-driver/src/infra/solver/dto/solution.rs (hunk 10 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/solution.rs]: Changed in 973e7135cebbd3a197fe23165029c990bfcdfeb0 so that only SDK types are aliased (got the `Sdk*` prefix)

File: crates/solana-driver/src/domain/solution.rs (hunk 9 above)
Comment (tilacog) [reply crates/solana-driver/src/domain/solution.rs]: Dropped in 0bd611466

File: crates/solana-driver/src/infra/solver/dto/auction.rs (hunk 8 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/auction.rs]: Addressed in 1a12165626d2ffa06fe31fd595ec85a5a0a9c79d

File: crates/solana-driver/src/infra/solver/dto/solution.rs (hunk 7 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/solution.rs:141]: Added a check + new error variant in 973e7135cebbd3a197fe23165029c990bfcdfeb0

File: crates/solana-driver/src/infra/solver/mod.rs (hunk 6 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs:72]: Addressed in bfdcbd6ebf5e58b45ca9d0fd6ad8feebfa0837e4

File: crates/solana-driver/src/infra/solver/dto/auction.rs (hunk 14 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/auction.rs:27]: Added the field in 00bbb8094.

I left a TODO in `solana-solvers` instead of implementing, to keep this PR focused.

File: crates/solana-driver/src/infra/solver/dto/solution.rs (hunk 11 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/solution.rs]: Addressed in 973e7135cebbd3a197fe23165029c990bfcdfeb0

File: crates/solana-driver/src/util/mod.rs (hunk 4 above)
Comment (tilacog) [reply crates/solana-driver/src/util/mod.rs:30]: This seems like a bigger change, so I'd rather handle it in a follow up PR.

(It involves extra IO and wiring the RPC value to the Solver strutct)

File: crates/solana-driver/src/infra/solver/dto/mod.rs (hunk 5 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/dto/mod.rs]: Added tests back in f7bf3e90836584173b93ee343fca9d62d693e7dc

File: crates/solana-driver/src/util/mod.rs (hunk 4 above)
Comment (squadgazzz) [reply crates/solana-driver/src/util/mod.rs:30]: @tilacog , would it be enough to loudly reject token 2022 for the demo or just adding a TODO? We are moving extremely slow with the driver implementation, so we have to avoid adding more functionality right now.

File: crates/solana-driver/src/infra/solver/mod.rs [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    },
+    /// The solver returned a response the driver could not interpret.
+    #[error("bad solver response: {0}")]
+    BadResponse(#[from] dto::solution::Error),
+    /// The request body could not be serialized.
+    #[error("JSON serialization error: {0}")]
+    Serialize(#[from] serde_json::Error),
+    /// The auction deadline passed before a solve request could be sent.
+    #[error("auction deadline exceeded")]
+    DeadlineExceeded,
```
Comment (squadgazzz) [inline crates/solana-driver/src/infra/solver/mod.rs:120]: `DeadlineExceeded` is likely a new error contract. Please add a test only if it is easy to implement, otherwise, add a TODO: a small test with a solve with a past deadline, so the caller gets `Err(DeadlineExceeded)` rather than an empty success.

**Review by squadgazzz — APPROVED** [review]
LGTM

File: crates/solana-driver/src/util/mod.rs (hunk 4 above)
Comment (tilacog) [reply crates/solana-driver/src/util/mod.rs:30]: Makes sense. The settlement contract code actually rejects token-2022, so we are not falling behind on that front.

Added a TODO comment in 533dc4a5cc9fa5034521fb843d452ade36621cdf.

File: crates/solana-driver/src/infra/solver/mod.rs (hunk 18 above)
Comment (tilacog) [reply crates/solana-driver/src/infra/solver/mod.rs:120]: Covered that in a test in 3690b1a70b3a6414ef7469ab1d9077f65adf43a6
